libmscore/figuredbass.cpp

```diff
@@ -1770,6 +1773,256 @@ FiguredBass* Score::addFiguredBass()
+      
+void FiguredBass::realizeFiguredBass(Score *score) {
+      Q_ASSERT(score);
 
+      for (Measure* measure = score->firstMeasure(); measure; measure = measure->nextMeasure()) {
+            realizeFiguredBass(measure);
+      }
+}
+      
+void FiguredBass::realizeFiguredBass(Measure *measure) {
+      Q_ASSERT(measure);
+      const Segment::Type st = Segment::Type::ChordRest;
+      for (Segment* segment = measure->first(st); segment; segment = segment->next(st)) {
+            realizeFiguredBass(segment);
+      }
+}
+      
+void FiguredBass::realizeFiguredBass(Segment *segment) {
+      Q_ASSERT(segment);
+      int etrack = segment->score()->nstaves() * VOICES;
+
+      for (int track = 0; track < etrack; ++track) {
+            Element *element = segment->element(track);
+            if ( element == 0)
+                  continue;
+            if ( element->type() != Element::Type::CHORD)
+                  continue;
+            {
+                  vector<int> harmonys;
+                  Chord *chord = static_cast<Chord*>(element);
+                  if ( calcFiguredBass(chord, harmonys) ) { // appends entries into harmonys corresponding musical harmony pitches
+                        for (int pitch : harmonys) {
+                              NoteVal *nval = new NoteVal(pitch);
+                              chord->score()->addNote(chord,*nval);
+                        }
+                  } else {
+                        //qDebug("no figured bass on chord on track %d", chord->track());
+                  }
+            }
+      }
+}
+      
+// populate the given harmonys vector with the pitches of the harmony notes implied by the figured base on the given chord
+bool FiguredBass::calcFiguredBass(Chord *chord, vector<int> & harmonys) {
+      Q_ASSERT(harmonys.size() == 0);
+      Segment *segment = chord->segment();
+      int octave = 0; // 0, or 12 if we want the harmony 1 octave up
+      FiguredBassItem::Modifier none = FiguredBassItem::Modifier::NONE;
+      if (find_if(segment->annotations().cbegin(),
+                  segment->annotations().cend(),
+                  [chord](Element *e) {
+                        return ( e->type() == Element::Type::FIGURED_BASS
+                                && e->track() == chord->track());
+                  }) == segment->annotations().cend())
+      {
+             return false; // no figured base to annotate
+      }
+      
+      if ( chord->notes().size() == 0 ) {
+            qDebug("no notes on chord");
+            return false; // no notes in the chord, not sure if this is possible?
+      }
+      if ( chord->notes().size() != 1 ) {
+            qDebug("measure #%d, cannot render Figured Base on a multi-note chord", segment->measure()->no());
+            return false;
+      }
+      Note *note = chord->notes()[0];
+
+      auto mod7 = [] (int n) {
+            n = n%7;
+            while ( n < 0)
+                  n += 7;
+            return n;
+      };
```

Comment:
What I can suggest is that sometime before the final commit, perhaps many times,  I’ll ask for your (collective) help
refactoring the code into a way which seems more native C++.  There may be times when removing a function will
make the code more obscure.  In this case you (the developers) can help me make things more clear.   How’s that sound?

Jim


> On 25 May 2015, at 16:09, Marc Sabatella <notifications@github.com> wrote:
> 
> In libmscore/figuredbass.cpp <https://github.com/musescore/MuseScore/pull/2015#discussion_r30981537>:
> 
> > +      if ( chord->notes().size() == 0 ) {
> > +            qDebug("no notes on chord");
> > +            return false; // no notes in the chord, not sure if this is possible?
> > +      }
> > +      if ( chord->notes().size() != 1 ) {
> > +            qDebug("measure #%d, cannot render Figured Base on a multi-note chord", segment->measure()->no());
> > +            return false;
> > +      }
> > +      Note *note = chord->notes()[0];
> > +
> > +      auto mod7 = [] (int n) {
> > +            n = n%7;
> > +            while ( n < 0)
> > +                  n += 7;
> > +            return n;
> > +      };
> My two cents - and there is no flaming implied: The reason to avoid these constructs is simple: C++ didn't support them for many years, experienced C++ programmers are accustomed to doing things in different ways, with millions of lines of code written and understand that doesn't use them, so .  introducing a new and rarely encountered paradigm into one specific area of the MuseScore source just makes it harder for experienced C++ programmers - and in particular, C++ programmers familiar with the MuseScore source - to understand that section of the code.  There's a reason we have coding guidelines to enforce a similar style across the application, and while guidelines might not be explicit about it, it seems obvious to me this applies just as much if not more to semantic style as to use of whitespace and capitalization. So while local functions and free passing about of function pointers may seem comfortable and familiar and easy to understand to someone who has been using that style of programming a long time, it just isn't part of C++ culture to use them for things that seem, to us, more inherently procedural.  If there is a specific *reason* to use them - to accomplish something not easily accomplished another way - that's great.  But as it is, it just makes C++ people scratch their hears wondering what unusual thing this section of code is trying to accomplish that meant it had to resort to this highly unusual structure. — Marc Sabatella
> … <x-msg://19/#>
> —
> Reply to this email directly or view it on GitHub.
> 
>  <https://github.com/musescore/MuseScore/pull/2015#discussion_r30972851> <https://github.com/lasconic> <https://github.com/lasconic> <https://github.com/musescore/MuseScore/pull/2015/files#r30972851> <https://github.com/musescore/MuseScore/pull/2015/files#r30979424>
> —
> Reply to this email directly or view it on GitHub <https://github.com/musescore/MuseScore/pull/2015/files#r30981537>.
>